### PR TITLE
make approval voting work on a small testnet

### LINKED ISSUE #2411: Approval Checking: Make it work for low-node testnets
In the import pipeline, we should automatically mark blocks as approved if the total amount of validators minus the amount of validators in the backing group is less than the number of needed approvals. This will require a small tweak to the API and call-site of `approval_db::v1::add_block_entry`.

Currently, testnets with a small number of nodes which have only a single parachain won't function as members of the backing group are prohibited from doing approval checks on the block.

### PR DESCRIPTION
**Pull request by ordian**
Fixes #2411

- [x] fix tests
- [x] add new test

### CODE REVIEW

**Comment by rphmeier** [comment]
Should rebase out all commits before this PR

File: node/core/approval-voting/src/import.rs [hunk 1]
```diff
@@ -605,7 +605,7 @@ pub(crate) async fn handle_new_head(
 				.copied()
 				.unwrap_or(0);
 			let needed_approvals = usize::try_from(needed_approvals).expect("usize is at least u32");
-			if needed_approvals + backing_group_size < n_validators {
+			if n_validators.saturating_sub(backing_group_size) < needed_approvals {
```
Comment (rphmeier) [inline node/core/approval-voting/src/import.rs:608]: Maybe should be `if needed_approvals == 0 || ... `?

File: node/core/approval-voting/src/import.rs [hunk 2]
```diff
@@ -599,16 +599,33 @@ pub(crate) async fn handle_new_head(
...
-				.copied()
-				.unwrap_or(0);
-			let needed_approvals = usize::try_from(needed_approvals).expect("usize is at least u32");
-			if n_validators.saturating_sub(backing_group_size) < needed_approvals {
-				approved_bitfield.set(i, true);
+		let approved_bitfield = {
+			if needed_approvals == 0 {
+				tracing::info!(
+					target: LOG_TARGET,
+					"Insta-approving approving all candidates",
```
Comment (rphmeier) [inline node/core/approval-voting/src/import.rs:606]: This will be really noisy :)

File: node/core/approval-voting/src/import.rs [hunk 3]
```diff
@@ -599,16 +599,33 @@ pub(crate) async fn handle_new_head(
...
+				let mut result = bitvec::bitvec![BitOrderLsb0, u8; 0; num_candidates];
+				for (i, &(hash, _, _, backing_group)) in included_candidates.iter().enumerate() {
+					let backing_group_size = validator_group_lens.get(backing_group.0 as usize)
+						.copied()
+						.unwrap_or(0);
+					let needed_approvals = usize::try_from(needed_approvals).expect("usize is at least u32");
+					if n_validators.saturating_sub(backing_group_size) < needed_approvals {
+						tracing::info!(
+							target: LOG_TARGET,
+							"Insta-approving candidate {:?} as the number of validators is too low",
```
Comment (rphmeier) [inline node/core/approval-voting/src/import.rs:619]: As will this

File: node/core/approval-voting/src/import.rs (hunk 2 above)
Comment (ordian) [reply node/core/approval-voting/src/import.rs:606]: this is per block, not candidate

File: node/core/approval-voting/src/import.rs (hunk 3 above)
Comment (ordian) [reply node/core/approval-voting/src/import.rs:619]: remove it or downgrade to debug? we will probably enable this log target anyway

File: node/core/approval-voting/src/import.rs (hunk 2 above)
Comment (rphmeier) [reply node/core/approval-voting/src/import.rs]: Yeah, still, that's a lot of noise. Debug is fine though!

**Review by rphmeier — APPROVED** [review]
Looks good! Needs test

**Review by rphmeier — APPROVED** [review]

**Pull request merged**